gtsfm/utils/features.py

```diff
@@ -47,3 +47,65 @@ def normalize_coordinates(coordinates: np.ndarray, intrinsics: Cal3Bundler) -> n
     """
 
     return np.vstack([intrinsics.calibrate(x[:2].reshape(2, 1)) for x in coordinates])
+
+
+def convert_to_homogenous_coordinates(
+    non_homogenous_coordinates: np.ndarray,
+) -> np.ndarray:
+    """Convert coordinates to homogenous system (by appending a column of ones).
+
+    Args:
+        non_homogenous_coordinates: 2d non-homogenous coordinates, of shape Nx2.
+
+    Returns:
+        2d homogenous coordinates, of shape Nx3.
+
+    Raises:
+        TypeError: if input is not 2 dimensional.
+    """
+    if non_homogenous_coordinates is None or non_homogenous_coordinates.size == 0:
+        return non_homogenous_coordinates
+
+    if non_homogenous_coordinates.shape[1] != 2:
+        raise TypeError("Input should be 2D")
+
+    return np.hstack(
+        (
+            non_homogenous_coordinates,
+            np.ones((non_homogenous_coordinates.shape[0], 1)),
+        )
+    )
+
+
+def convert_to_epipolar_lines(normalized_coordinates_i1: np.ndarray, i2Ei1: EssentialMatrix) -> np.array:
+    """Convert coordinates to epipolar lines.
+
+    Args:
+        normalized_coordinates_i1: normalized coordinates in i1, of shape Nx2.
+        i2Ei1: essential matrix.
+
+    Returns:
+        Corr. epipolar lines in i2, of shape Nx3
+    """
+    if normalized_coordinates_i1 is None or normalized_coordinates_i1.size == 0:
+        return normalized_coordinates_i1
+
+    epipolar_lines = convert_to_homogenous_coordinates(normalized_coordinates_i1) @ i2Ei1.matrix().T
+
+    return epipolar_lines
+
+
+def compute_point_line_distances(points: np.ndarray, lines: np.ndarray) -> np.ndarray:
+    """Computes the distance of a point from a line in 2D. The function processed multiple inputs independently in a
+    vectorized fashion.
+
+    Args:
+        points: non-homogenous 2D points, of shape Nx2.
+        lines: lines, of shape Nx3.
+
+    Returns:
+        Point-line distance for each row, of shape N.
+    """
+    line_norms = np.linalg.norm(lines[:, :2], axis=1)
+
+    return np.abs(np.einsum("ij,ij->i", convert_to_homogenous_coordinates(points), lines)) / line_norms
```

Comment:
i personally dont find einsum to be too readable, so I'd prefer not to use it

Reply:
Here I want to do row-wise dot product. If I do not use einsum, I have to use `np.sum(np.multiply(a, b), axis=1)`. I thought this alternative is not elegant.

Reply:
can we write it with a matrix multiplication?

Reply:
bmm?

Reply:
Does numpy has bmm?

Reply:
matmul can do something similar. But i personally think `np.sum(np.multiply(a, b), axis=1)` is clear and good

Reply:
resolved